src/client/Client.cc

```diff
@@ -7203,10 +7204,11 @@ int Client::mkdir(const char *relpath, mode_t mode, const UserPerm& perm, std::s
   path.pop_dentry();
   InodeRef dir;
 
-  std::scoped_lock lock(client_lock);
   int r = path_walk(path, &dir, perm);
   if (r < 0)
     return r;
+
+  std::scoped_lock lock(client_lock);
   if (cct->_conf->client_permissions) {
```

Comment:
Commenting here because github doesn't allow comments against the changelog. The patch makes sense, but the commit log is bogus. This is not a "cleanup" per-se, but rather a lock pushdown, where we're reducing the footprint of the client_lock by having a lower-level function take it itself.

Aside from that, we should note that you're increasing the mutex thrashing here. Previously, we'd take the client lock once and just release it when Client::mkdir returns. Now you take it in path_walk, release it when it returns, and take it again here for the rest of mkdir. That's generally less efficient than it was before. If this is an interim change that gets cleaned up later then that's ok. Otherwise, we'll need to weigh whether this change is worthwhile.

Reply:
This is an interim change, and preparing for the inode lock. Without this we need to release the `client_lock` in the `path_walk` to take the  `inode_lock` and take the client_lock again, like we do in the kclient:

```
client_lock.unlock();
std::scoped_lock il{in->inode_lock};
client_lock.lock();
```

Reply:
This is an interim change, and preparing for the inode lock. Without this we need to release the `client_lock` in the `path_walk` to take the  `inode_lock` and take the client_lock again, like we do in the kclient:

```
client_lock.unlock();
std::scoped_lock il{in->inode_lock};
client_lock.lock();
```

Reply:
I add some logs and have found that ususally it will take 80ns in average to return from acquiring the mutex locks.

Without the inode lock patches, the `client_lock` ususally will wait 15ms in average if it needs to wait, with this PR it will reduce to 8ms in average.

So the thrashing should be acceptable ?

This is not that precise, I just add some logs randomly in the code to get the results.

Reply:
Cool, this is what I can do to split the patch into small ones, or all something like this must be squashed into the inode lock one.